### PR TITLE
Add forge test for Mailbox

### PR DESCRIPTION
**Pull request by yorhodes**
### Description

Updates v3 contracts with forge tests

### Drive-by changes

Remove mailbox hardhat tests

### Backward compatibility

Yes

### Testing

Unit Tests

### CODE REVIEW

File: solidity/test/Mailbox.t.sol [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+                metadata,
+                overrideHook
+            );
+            assertEq(mailbox.latestDispatchedId(), id);
+            nonce = mailbox.nonce();
+            assertEq(nonce, i + 3);
+        }
+    }
+
+    function test_100dispatch_withMerkleTreeHook(bytes calldata body) public {
```
Comment (aroralanuk) [inline solidity/test/Mailbox.t.sol:294]: can fuzz 100 too?

File: solidity/test/Mailbox.t.sol [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+            remoteDomain,
+            recipientb32,
+            msg.sender.addressToBytes32(),
+            body
+        );
+        bytes32 id = keccak256(message);
+        assertEq(mailbox.delivered(id), false);
+        expectProcess(metadata, message, body, value);
+        mailbox.process{value: value}(metadata, message);
+        assertEq(mailbox.delivered(id), true);
```
Comment (aroralanuk) [inline solidity/test/Mailbox.t.sol:355]: expectProcessed(id)?

File: solidity/test/hyperlaneConnectionClient.test.ts [hunk 3]
```diff
@@ -27,8 +27,9 @@ describe('HyperlaneConnectionClient', async () => {
   beforeEach(async () => {
```
Comment (aroralanuk) [inline solidity/test/hyperlaneConnectionClient.test.ts:27]: bearish

File: solidity/test/mockMailbox.test.ts [hunk 4]
```diff
@@ -23,7 +23,7 @@ describe('MockMailbox', function () {
 
     const body = ethers.utils.toUtf8Bytes('This is a test message');
 
-    await originMailbox.dispatch(
+    await originMailbox['dispatch(uint32,bytes32,bytes)'](
```
Comment (aroralanuk) [inline solidity/test/mockMailbox.test.ts:26]: bearish

**Review by aroralanuk — APPROVED** [review]

File: solidity/test/Mailbox.t.sol [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+        );
+        assertEq(quote, defaultFee + requiredFee);
+
+        expectHookQuote(requiredHook, metadata, message);
+        expectHookQuote(defaultHook, metadata, message);
+        quote = mailbox.quoteDispatch(
+            remoteDomain,
+            address(recipient).addressToBytes32(),
+            body,
+            metadata
```
[the remark this answers is not in the document]
Comment (yorhodes) [reply solidity/test/Mailbox.t.sol:207]: im not testing the igp hook here

File: solidity/test/Mailbox.t.sol (hunk 1 above)
Comment (yorhodes) [reply solidity/test/Mailbox.t.sol:294]: for sure
this was mostly for instrumenting gas

File: solidity/contracts/Router.sol [hunk 6]
```diff
@@ -193,31 +201,31 @@ abstract contract Router is HyperlaneConnectionClient, IMessageRecipient {
...
-        interchainGasPaymaster.payForGas{value: _gasPayment}(
-            _messageId,
-            _destinationDomain,
+        // Ensure that destination chain has an enrolled router.
+        bytes32 _router = _mustHaveRemoteRouter(_destinationDomain);
+        bytes memory metadata = IGPMetadata.formatMetadata(
             _gasAmount,
             _gasPaymentRefundAddress
         );
+        _messageId = mailbox.dispatch{value: _gasPayment}(
```
[the remark this answers is not in the document]
Comment (yorhodes) [reply solidity/contracts/Router.sol:210]: good question
started a thread in discord

[the remark this answers is not in the document]
Comment (yorhodes) [reply solidity/contracts/Router.sol:210]: seems like we will punt for now

File: solidity/test/Mailbox.t.sol (hunk 1 above)
Comment (yorhodes) [reply solidity/test/Mailbox.t.sol:319]: added a comment

**Pull request merged**